BucketAccessRequest field update

copejon opened the pull request:
This PR replaces:

- `bucketAccessRequest.spec.bucket` with `bucketAccessRequest.spec.bucketRequest`.  The Bucket should be dereferenced through the user's BucketRequest.

- `bucketAccessRequest.spec.bucketAccessRequestName` field with `bucketAccessRequest.spec.bucketAccessName` to reflect the relationship between the user and controller side APIs.

Edit to include:
Notable changes include:

- simplified controller-gen, result is CRDs are genereated to config/crd/, rather than config/crd/bases
- set path kb tags to force k8s DNS formatted object names. Else crds are generated with metadata.name: CamelCaseObject  which fails
- manually added approval annotation.  this is a pain as it's obliterated every time the CRDs are regenerated.  Will require a small custom tool.
- and the actual stated purpose of the PR: brought the go types in line with the KEP definitions

copejon committed:
update bucket* APIs

change bucket -> bucketName

fixed typoed const name

fixed json tag

copejon commented:
Running code generator produces:
```
$ ./hack/update-codegen.sh
API rule violation: list_type_missing,github.com/container-object-storage-interface/api/apis/cosi.sigs.k8s.io/v1alpha1,BucketAccessClass,SupportedProtocols
API rule violation: list_type_missing,github.com/container-object-storage-interface/api/apis/cosi.sigs.k8s.io/v1alpha1,BucketClass,AdditionalPermittedNamespaces
API rule violation: list_type_missing,github.com/container-object-storage-interface/api/apis/cosi.sigs.k8s.io/v1alpha1,BucketClass,AnonymousAccessModes
API rule violation: list_type_missing,github.com/container-object-storage-interface/api/apis/cosi.sigs.k8s.io/v1alpha1,BucketClass,SupportedProtocols
API rule violation: list_type_missing,github.com/container-object-storage-interface/api/apis/cosi.sigs.k8s.io/v1alpha1,BucketSpec,PermittedNamespaces
API rule violation: list_type_missing,github.com/container-object-storage-interface/api/apis/cosi.sigs.k8s.io/v1alpha1,PolicyActions,Allow
API rule violation: list_type_missing,github.com/container-object-storage-interface/api/apis/cosi.sigs.k8s.io/v1alpha1,PolicyActions,Deny
API rule violation: names_match,github.com/container-object-storage-interface/api/apis/cosi.sigs.k8s.io/v1alpha1,BucketAccessRequestSpec,BucketRequestName
```

copejon commented:
dependent on #2 for API fixes

wlan0 commented on apis/cosi.sigs.k8s.io/v1alpha1/protocol_types.go:
```diff
@@ -5,7 +5,7 @@ type ProtocolSignature string
 const (
 	ProtocolSignatureS3    ProtocolSignature = "s3"
 	ProtocolSignatureAzure ProtocolSignature = "azure"
-	ProtcolSignatureGCS    ProtocolSignature = "gcs"
```
This is a nice find.

wlan0 commented:
Please remove the .DS_Store file and rebase. Otherwise LGTM

copejon committed:
updated go type APIs

moved genereated CRDs to config/crds simplify dir tree

added kubebuilder defaulter, validation tags

added api approval annotation

copejon committed:
refined generate command

copejon committed:
added path override to go types to force k8s adherent object names

copejon committed:
rm .DS_Store file

wlan0 commented:
LGTM
